src/api/mod.rs

```diff
@@ -1,68 +1,105 @@
 use std::collections::HashMap;
 
-use crate::utility::{URI, ENDPOINT_CREDITS, ZBError};
-
-pub struct ZeroBounce<'key> {
-    api_key: &'key str,
-    client: reqwest::blocking::Client
-}
-
-impl ZeroBounce<'_> {
-    pub fn new(api_key: &str, client: reqwest::blocking::Client) -> ZeroBounce {
-        ZeroBounce {api_key, client}
+use chrono::NaiveDate;
+use serde_json::from_str;
+
+use crate::{ZeroBounce, ZBUrlProvider};
+use crate::utility::{ZBError, ZBResult, structures::ApiUsage};
+
+impl ZeroBounce {
+    pub fn new(api_key: &str) -> ZeroBounce {
+        ZeroBounce {
+            api_key: api_key.to_string().clone(),
+            client: reqwest::blocking::Client::default(),
+            url_provider: ZBUrlProvider::default()
+        }
     }
 
-    fn get_credits_from_string(string_value: String) -> Result<i64, ZBError> {
-
-        let parsed_value = serde_json::from_str::<serde_json::Value>(string_value.as_ref())
-            .map_err(ZBError::JsonError)?;
-
-        parsed_value
+    fn get_credits_from_string(string_value: String) -> ZBResult<i64> {
+        from_str::<serde_json::Value>(string_value.as_ref())?
             .get("Credits")
             .and_then(serde_json::Value::as_str)
             .map(str::parse::<i64>)
             .ok_or(ZBError::explicit("credits value not in json"))?
             .map_err(ZBError::IntParseError)
     }
 
-    pub fn get_credits(&self) -> Result<i64, ZBError> {
-        
-        let url_string = URI.to_owned() + ENDPOINT_CREDITS;
-        
-        let mut query_args = HashMap::<&str, &str>::new();
-        query_args.insert("api_key", self.api_key);
+    pub fn get_credits(&self) -> ZBResult<i64> {
+        let url = self.url_provider.url_of(crate::utility::ENDPOINT_CREDITS);
+        let mut query_args = HashMap::<&str, String>::new();
 
-        let response = self.client.get(url_string)
+        query_args.insert("api_key", self.api_key.clone());
+
+        let response = self.client.get(url)
             .query(&query_args)
-            .send()
-            .map_err(ZBError::RequestError)?;
+            .send()?;
 
         let response_ok = response.status().is_success();
-
         let response_content = response
-            .text()
-            .map_err(ZBError::RequestError)?;
+            .text()?;
 
         if !response_ok {
             return Err(ZBError::explicit(response_content.as_str()));
         }
 
         Self::get_credits_from_string(response_content)
     }
+
+    pub fn get_api_usage(&self, start_date: NaiveDate, end_date:NaiveDate) -> ZBResult<ApiUsage> {
+        let url = self.url_provider.url_of(crate::utility::ENDPOINT_API_USAGE);
+
+        let mut query_args = HashMap::<&str, String>::new();
+        let start_date_str = start_date.format("%F").to_string();
+        let end_date_str = end_date.format("%F").to_string();
+
+        query_args.insert("api_key", self.api_key.clone());
```

Comment:
you can do this: 
![image](https://user-images.githubusercontent.com/108796024/230338656-e42c90b1-7eb6-4a5b-b789-a20df04df05a.png)

Reply:
Done that eventually, in the other PRs